encode/vp8enc.c

```diff
@@ -0,0 +1,1173 @@
+  picParam->ref_gf_frame = VA_INVALID_SURFACE;
+  picParam->ref_arf_frame = VA_INVALID_SURFACE;
+
+  /* always show it */
+  picParam->pic_flags.bits.show_frame = 1;
+
+  for (size_t i = 0; i < N_ELEMENTS(picParam->loop_filter_level); i++) {
+      picParam->loop_filter_level[i] = settings.loop_filter_level;
+  }
+
+  picParam->clamp_qindex_low = settings.clamp_qindex_low;
+  picParam->clamp_qindex_high = settings.clamp_qindex_high;
+
+}
+
+void vp8enc_set_refreshparameter_for_svct_2layers(VAEncPictureParameterBufferVP8 *picParam, int current_frame, int *is_golden_refreshed)
+{
+  //Pattern taken from libyami
+
+  picParam->ref_flags.bits.no_ref_arf = 1;
+
+  if(! *is_golden_refreshed)
+    picParam->ref_flags.bits.no_ref_gf = 1;
+
+  switch(current_frame % 2) {
+    case 0:
+      //Layer 0
+      picParam->pic_flags.bits.refresh_last = 1;
+      picParam->ref_flags.bits.no_ref_gf = 1;
+      picParam->ref_flags.bits.temporal_id = 0;
+      break;
+    case 1:
+      //Layer 1
+      picParam->pic_flags.bits.refresh_golden_frame = 1;
+      *is_golden_refreshed = 1;
+      picParam->ref_flags.bits.temporal_id = 1;
+      break;
+  }
+}
+
+void vp8enc_set_refreshparameter_for_svct_3layers(VAEncPictureParameterBufferVP8 *picParam, int current_frame,int *is_golden_refreshed)
+{
+  //Pattern taken from libyami - Note that the alternate frame is never referenced,
+  //this is because, libyami implementation suggests to be able to drop individual
+  //frames from Layer 2 on bad network connections
+  picParam->ref_flags.bits.no_ref_arf = 1;
+
+  if(! *is_golden_refreshed)
+    picParam->ref_flags.bits.no_ref_gf = 1;
+
+  switch(current_frame % 4) {
+    case 0:
+      //Layer 0
+      picParam->pic_flags.bits.refresh_last = 1;
+      picParam->ref_flags.bits.no_ref_gf = 1;
+      picParam->ref_flags.bits.temporal_id = 0;
+      break;
+    case 1:
+    case 3:
+      //Layer 2
+      picParam->pic_flags.bits.refresh_alternate_frame  = 1;
+      picParam->ref_flags.bits.temporal_id = 2;
+      break;
+    case 2:
+      //Layer 1
+      picParam->pic_flags.bits.refresh_golden_frame = 1;
+      *is_golden_refreshed = 1;
+      picParam->ref_flags.bits.temporal_id = 1;
+      break;
+  }
+}
+
+void vp8enc_reset_picture_parameter_references(VAEncPictureParameterBufferVP8 *picParam)
+{
+  picParam->ref_last_frame = VA_INVALID_SURFACE;
+  picParam->ref_gf_frame = VA_INVALID_SURFACE;
+  picParam->ref_arf_frame = VA_INVALID_SURFACE;
+  picParam->pic_flags.bits.refresh_last = 0;
+  picParam->pic_flags.bits.refresh_golden_frame = 0;
+  picParam->pic_flags.bits.refresh_alternate_frame = 0;
+  picParam->pic_flags.bits.copy_buffer_to_golden = 0;
+  picParam->pic_flags.bits.copy_buffer_to_alternate = 0;
+  picParam->ref_flags.bits.no_ref_last = 0;
+  picParam->ref_flags.bits.no_ref_gf = 0;
+  picParam->ref_flags.bits.no_ref_arf = 0;
+}
+
+void vp8enc_update_picture_parameter(int frame_type, int current_frame)
+{
+  static int is_golden_refreshed = 0;
```

Comment:
Could we remove this static int. it's not re-enterable. https://en.wikipedia.org/wiki/Reentrancy_(computing). Hard to debug a no re-enterable function

Reply:
I can make is_golden_refreshed global and maybe place it in vaapi_context

Reply:
actually I don't know if making the flag global actually improves the situation - the whole application is on multiple occasions non re-entrant. - Actually also global variables are not allowed for reentrancy - as stated in the wiki one of the rules is: "Reentrant code may not hold any static or global non-constant data."